statistics/statistics.go

```diff
@@ -21,329 +21,139 @@ package statistics
-	return *policy, nil
-
-}
-
-type byTimestamp []plugin.Metric
-
-func (sa byTimestamp) Len() int {
-	return len(sa)
-}
 
-func (sa byTimestamp) Less(i, j int) bool {
-	return sa[i].Timestamp.Before(sa[j].Timestamp)
-
-}
-func (sa byTimestamp) Swap(i, j int) {
-	sa[i], sa[j] = sa[j], sa[i]
-}
-
-func (p *Plugin) insertInToTimeBuffer(metric plugin.Metric, times []time.Time) []time.Time {
-	times[p.bufferIndex] = metric.Timestamp
-	return times
-}
-
-func (p *Plugin) getTimes(times []time.Time) (time.Time, time.Time) {
-	if p.bufferCurSize == p.bufferMaxSize && p.bufferIndex != p.bufferMaxSize-1 {
-		return times[p.bufferIndex+1], times[p.bufferIndex]
+	if errString != "" {
+		err = fmt.Errorf(errString)
 	}
-	return times[0], times[p.bufferIndex]
+	return
 }
 
-// Process processes the data, inputs the data into this' buffer and calls the descriptive statistics method
-func (p *Plugin) Process(metrics []plugin.Metric, cfg plugin.Config) ([]plugin.Metric, error) {
-	/* f, err := os.OpenFile("/tmp/statisticErr.txt", os.O_WRONLY|os.O_CREATE, 0755)
-	if err != nil {
-		log.Warn("File reading error.")
-	}
-
-	log.SetOutput(f) */
-
-	bufsize, err := cfg.GetInt("SlidingWindowLength")
-	p.bufferMaxSize = int(bufsize)
-	if err != nil {
-		return nil, err
-	}
-
-	var results []plugin.Metric
-	times := make([]time.Time, p.bufferMaxSize)
-
-	metricNamespace := make(map[string][]plugin.Metric)
-	for _, metric := range metrics {
-		//Populates the metricNamespace map so that the statistics are ran on metrics that share the same namespace.
-		ns := metric.Namespace.String()
-		if plugins, ok := metricNamespace[ns]; ok {
-			plugins = append(plugins, metric)
-			metricNamespace[ns] = plugins
-		} else {
-			metricNamespace[ns] = []plugin.Metric{metric}
-		}
-	}
-
-	for k, v := range metricNamespace {
-		var startTime time.Time
-		var stopTime time.Time
-		unit := v[0].Unit
-
-		sort.Sort(byTimestamp(v))
-		for _, metric := range v {
-			times = p.insertInToTimeBuffer(metric, times)
-			startTime, stopTime = p.getTimes(times)
-			p.insertInToBuffer(metric.Data, metric.Namespace.String())
-			p.updateCounters()
-
-			if p.bufferCurSize < p.bufferMaxSize {
-				stats, err := p.calculateStats(p.buffer[k][0:p.bufferCurSize], startTime, stopTime, k, unit)
-				if err != nil {
-					log.Warnf("Error occured in calculating Statistics: %s", err)
-					return nil, err
-				}
-				results = append(results, stats...)
-			} else if p.bufferCurSize == p.bufferMaxSize {
-				stats, err := p.calculateStats(p.buffer[k], startTime, stopTime, k, unit)
-				if err != nil {
-					log.Warnf("Error occured in calculating Statistics: %s", err)
-					return nil, err
-				}
-				results = append(results, stats...)
-			}
-		}
-	}
-
-	return results, nil
+// converts data to float64 type
+func dataToFloat64(data interface{}) (float64, error) {
+	var value float64
+	switch v := data.(type) {
+	case int:
```

Comment:
As nil is not a type ([see here](https://play.golang.org/p/dUdE5C8qeC)), we should catch the nil value before the switch, is it correct?